### PR TITLE
Provide an alternative for --set and/or --set-string to take a value literally

### LINKED ISSUE #4030: Provide an option or alternative for --set and/or --set-string to take a value literally
I am experiencing an escaping problem with Helm. For instance, if I want to set a password in a chart, I run into the problem that `\` characters are removed.

Example:

1. Take a chart that sets a k8s secret where the value of the secret is base64 encoded by Helm (b64enc)
1. Run `helm upgrade --install --set password=Passw\Ord! somechart`
1. The resulting base64 encoded string decodes to `PasswOrd!` instead of `Passw\Ord!`

I looked into possible solutions on my end and did the following:
1. Singly escaping the `\` resulting in `Passw\\0rd!` does not give the desired output, both `\` are removed, resulting in `PasswOrd!`.
1. Wrapping with either ticks (`'`) or quotes (`"`) results `\` being removed, resulting in `PasswOrd!`.

What did work though is escaping the backslash two times `Passw\\\Ord!`, which results in `Passw\Ord!`.

# Expected behaviour
I would expect that `\\` would result in `\` instead of both being removed. 

# Desired outcome
It should be possible for passwords to contain special characters, such as `\`. 

# Possible solution:
Wrapped variable values in the `--set` argument should be taken literally (e.g. `--set password='Passw\Ord!'` should result in `Passw\Ord!`.

Output of `helm version`:
2.8.2

Output of `kubectl version`:
Client: 1.10.1
Server: 1.9.6

Edit: updated character naming to be more consistent on what is meant

### PR DESCRIPTION
**Pull request by pscheid92**
Currently, just an experiment/spike.\
Going to fill PR with more meaningful information and rebase/reword commits later.

This – hopefully – resolves #4030.

Signed-off-by: Patrick Scheid <p.scheid92@gmail.com>

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by bacongobbler** [comment]
Thanks you for taking this on!

I see that CircleCI is failing and has been in a "work-in-progress" state for 3 weeks. What's the status here?

Thanks!

**Comment by pscheid92** [comment]
Hello, unfortunately I had a lot going on in the last weeks and had problems finding time to work on it.\
Currently, things are smoothing down, and I hope I can continue working on this soon.

**Comment by buckner** [comment]
Hi - is there any traction on this? --set-literal would be immensely helpful to have.

**Comment by pscheid92** [comment]
Hello @bacongobbler I just want to make sure I haven't missed something. Is there any next step I have to take to bring this forward?

**Comment by Chainsaw-does-brr** [comment]
@bacongobbler , the commas gather their pitchforks:
,,,,,,,,,,,,,,,, , , , ,,  , , ,, ,,,,,,,,, , , ,, , ,, , ,,,,,,,,,,, ,, , , , , ,, , ,,,,,,,,,,,,
, ,, , ,,,,,,,,,,,,, ,, , , ,,,,,,,,,, , , ,,,,,, ,,,, ,,  , , ,, ,,,,,,,,, , , ,, , ,, , ,,,,, ,
We need to exclude them from parsing immediately!

**Comment by lsowen** [comment]
Looks like this would also address https://github.com/helm/helm/issues/5618

**Review by LukeIGS — APPROVED** [review]
Unit tests pass locally, seems to adhere the code styles set forth by the project to the best of what i can tell.

**Comment by f1tzpatrick** [comment]
Hello @marckhouzam or @mattfarina, when you have a chance, would you please review this PR? The functionality described in #4030 would be nice to have and this looks like it should add it.

**Review by bacongobbler — APPROVED** [review]

**Comment by bacongobbler** [comment]
Needs another maintainer’s approval as per our contributing policy. Thanks for the contribution.

**Comment by Alceatraz** [comment]
--set-literal, That gone *literal* save the world 
```bash
# Say good bye to those nightmare
--set SPRING_PROFILE='elk\,foo-prod\,bar-prod\,foobar-prod'
--set FOO_BAR='\{\"foo\":\"bar\"\}'
```

**Comment by Alceatraz** [comment]
Will this RP merge in 3.9.x?

**Comment by zarko-a** [comment]
Any chance this will get merged soon? 

We encountered the same issue and this would help us immensely.

**Comment by bacongobbler** [comment]
I've added it to the 3.9 milestone, which is scheduled for May 2022.

**Comment by hickeyma** [comment]
@pscheid92 Did you mind rebasing?

**Comment by pscheid92** [comment]
Hey @hickeyma I rebased my changes on top of the current upstream main branch (and resolved minor merge conflicts).

But it looks like something is wrong with the CircleCI Pipeline 🤔 
> Could not find a usable config.yml, you may have revoked the CircleCI OAuth app.
Please sign out of CircleCI and log back in with your VCS before triggering a new pipeline.

**Comment by pscheid92** [comment]
Hey @hickeyma @bacongobbler. Can I do something to help here? It still seems like the CircleCI Pipeline is somehow broken 🤔

**Comment by rogaha** [comment]
any updates?

**Comment by mattfarina** [comment]
I just fixed the merge conflict which is triggering  CI

File: pkg/strvals/literal_parser.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+			}
+
+			// recurse on sub-tree with remaining data
+			err := t.key(inner)
+			if len(inner) == 0 {
+				return errors.Errorf("key map %q has no value", string(key))
+			}
+			set(data, string(key), inner)
+			return err
+
```
Comment (hickeyma) [inline pkg/strvals/literal_parser.go:130]: There is a potential stack overflow if the nested levels are not checked. Do you mind adding a check for the key similar what is done in `parser.go`: https://github.com/helm/helm/blob/main/pkg/strvals/parser.go#L267-L288

**Review by hickeyma — CHANGES_REQUESTED** [review]
Thank you for pushing the PR @pscheid92 and your perservence. It is looking good except for a check on the levels as specified inline. It would be worth also adding an unit test similar to `TestParseSetNestedLevels` in TestParseSetNestedLevels

File: pkg/strvals/literal_parser.go (hunk 1 above)
Comment (pscheid92) [reply pkg/strvals/literal_parser.go]: Hey @hickeyma, I just pushed a version with stack overflow protection. And also included a unit test for it.

**Comment by pscheid92** [comment]
Thank you very much @hickeyma. I just pushed a version with stack overflow protection and the requested test case. Good catch!

File: pkg/strvals/literal_parser.go (hunk 1 above)
Comment (hickeyma) [reply pkg/strvals/literal_parser.go]: Great, thanks

**Review by hickeyma — APPROVED** [review]
I tested this manually and it fixes passing strings like `Passw\Ord!`.  I used `--set-literal passwd="Passw\\Ord\!"` flag and it output `Passw\Ord!`.

Nice feature, thanks @pscheid92

**Comment by hickeyma** [comment]
@mattfarina Are you happy for this to be merged?

**Comment by Alceatraz** [comment]
Finily

**Review by mattfarina — APPROVED** [review]

**Pull request merged**